File: compiler/bench/caesar.fs
\ expected: 3734
\ Caesar cipher benchmark - encode string, sum ASCII values, 50000 times

create teststr 64 allot
create outstr 64 allot

: init-str ( -- )
  s" The quick brown fox jumps over lazy dog" drop teststr 39 cmove ;

: caesar-char ( c shift -- c' )
  over [char] A [char] Z 1+ within if
    swap [char] A - + 26 mod [char] A + exit
  then
  over [char] a [char] z 1+ within if
    swap [char] a - + 26 mod [char] a + exit
  then
  drop ;

: caesar ( addr len shift -- )
  -rot 0 do
    2dup i + c@ swap caesar-char
    outstr i + c!
  loop 2drop ;

: sum-out ( len -- sum )
  0 swap 0 do outstr i + c@ + loop ;

: main ( -- )
  init-str
  0
  50000 0 do
    drop
    teststr 39 5 caesar
    39 sum-out
  loop
  . cr ;

File: compiler/bench/caesar.c
// expected: 3734
// Caesar cipher benchmark - encode string, sum ASCII values, 50000 times

#include <stdio.h>
#include <string.h>

char teststr[64];
char outstr[64];

void init_str(void) {
    strcpy(teststr, "The quick brown fox jumps over lazy dog");
}

char caesar_char(char c, int shift) {
    if (c >= 'A' && c <= 'Z') {
        return 'A' + (c - 'A' + shift) % 26;
    }
    if (c >= 'a' && c <= 'z') {
        return 'a' + (c - 'a' + shift) % 26;
    }
    return c;
}

void caesar(const char *src, int len, int shift) {
    for (int i = 0; i < len; i++) {
        outstr[i] = caesar_char(src[i], shift);
    }
}

int sum_out(int len) {
    int sum = 0;
    for (int i = 0; i < len; i++) {
        sum += (unsigned char)outstr[i];
    }
    return sum;
}

int main(void) {
    init_str();
    int result = 0;
    for (int i = 0; i < 50000; i++) {
        caesar(teststr, 39, 5);
        result = sum_out(39);
    }
    printf("%d\n", result);
    return 0;
}
